File: frontend/src/service/categoryService.ts
import api from "../config/axios";
import type { ApiResponse } from "../types/ApiResponse";
import type { PageResponse } from "../types/PageResponse";
import type {
  CategoryResponse,
  CategoryRequest,
  UpdateCategoryRequest,
} from "../types/category";

class CategoryService {
  async getAllCategories(
    page: number = 1,
    size: number = 10,
    keyword?: string,
    from?: string,
    to?: string,
  ) {
    const params = new URLSearchParams();
    params.append("page", page.toString());
    params.append("size", size.toString());

    if (keyword) params.append("keyword", keyword);
    if (from) params.append("from", from);
    if (to) params.append("to", to);

    const response = await api.get<ApiResponse<PageResponse<CategoryResponse>>>(
      `/categories?${params.toString()}`,
    );
    return response.data;
  }

  async getCategoryById(categoryId: string) {
    const response = await api.get<ApiResponse<CategoryResponse>>(
      `/categories/${categoryId}`,
    );
    return response.data;
  }

  async createCategory(request: CategoryRequest) {
    const response = await api.post<ApiResponse<CategoryResponse>>(
      "/categories",
      request,
    );
    return response.data;
  }

  async updateCategory(categoryId: string, request: UpdateCategoryRequest) {
    const response = await api.put<ApiResponse<CategoryResponse>>(
      `/categories/${categoryId}`,
      request,
    );
    return response.data;
  }

  async deleteCategory(categoryId: string) {
    const response = await api.delete<ApiResponse<void>>(
      `/categories/${categoryId}`,
    );
    return response.data;
  }
}

export default new CategoryService();

File: mobile/lib/services/category_service.dart
import '../models/page_response.dart';
import '../models/category.dart';
import '../config/api_client.dart';

class CategoryService {
  Future<PageResponse<CategoryResponse>> getAllCategories({
    int page = 1,
    int size = 10,
    String? keyword,
    String? from,
    String? to,
  }) async {
    final params = <String, dynamic>{
      'page': page,
      'size': size,
      if (keyword != null && keyword.isNotEmpty) 'keyword': keyword,
      if (from != null && from.isNotEmpty) 'from': from,
      if (to != null && to.isNotEmpty) 'to': to,
    };

    final res = await apiClient.get('/categories', queryParameters: params);

    return PageResponse<CategoryResponse>.fromJson(
      res.data['result'],
      (json) => CategoryResponse.fromJson(json),
    );
  }

  Future<CategoryResponse> getCategoryById(String categoryId) async {
    final res = await apiClient.get('/categories/$categoryId');
    return CategoryResponse.fromJson(res.data['result']);
  }

  Future<CategoryResponse> createCategory(Map<String, dynamic> request) async {
    final res = await apiClient.post('/categories', data: request);
    return CategoryResponse.fromJson(res.data['result']);
  }

  Future<CategoryResponse> updateCategory(
    String categoryId,
    Map<String, dynamic> request,
  ) async {
    final res = await apiClient.put('/categories/$categoryId', data: request);
    return CategoryResponse.fromJson(res.data['result']);
  }

  Future<void> deleteCategory(String categoryId) async {
    await apiClient.delete('/categories/$categoryId');
  }
}

final categoryService = CategoryService();
